feat: port real number parsing

MMZK1526 opened the pull request:
- [x] error configs
- [x] unsigned float (`Rational`) parser
- [x] signed float (`Rational`) parser

MMZK1526 committed:
feat: port error configs related to real number parsing

j-mie6 commented:
Can you shift the `cache` to `v4` in the CI?

j-mie6 committed:
Merge branch 'main' into feat/port-real

MMZK1526 committed:
Merge branch 'main' into feat/port-real

MMZK1526 commented:
For `BigDecimal` from Scala, can I port it with [`Data.BigDecimal`](https://hackage.haskell.org/package/HasBigDecimal-0.2.0.0/docs/Data-BigDecimal.html)? If I understand correctly, both are real numbers with a fixed arbitrary precision after the decimal points.

j-mie6 commented:
> For `BigDecimal` from Scala, can I port it with [`Data.BigDecimal`](https://hackage.haskell.org/package/HasBigDecimal-0.2.0.0/docs/Data-BigDecimal.html)? If I understand correctly, both are real numbers with a fixed arbitrary precision after the decimal points.

I would go with `Rational`: https://wiki.haskell.org/Rational. The way the float parsers work is by adding fractional powers of 2/8/10/16, so ultimately they can always be precisely represented with a `Rational`: there are no recurrences or irrationals that can be encoded in a finite number of digits

MMZK1526 committed:
feat: introduce `FloatingParsers` type

MMZK1526 commented on src/Text/Gigaparsec/Internal/Token/Numeric.hs:
```diff
@@ -371,23 +371,35 @@ mkSigned NumericDesc{..} !unsigned !err = IntegerParsers {
 
-mkUnsignedFloating :: NumericDesc -> IntegerParsers CanHoldUnsigned -> GenericNumeric -> FloatingParsers
-mkUnsignedFloating NumericDesc{..} nat gen = FloatingParsers {}
+-- mkUnsignedFloating :: NumericDesc -> IntegerParsers CanHoldUnsigned -> GenericNumeric -> ErrorConfig -> FloatingParsers
+-- mkUnsignedFloating NumericDesc{..} nat gen err = FloatingParsers {}
```
To implement this, should I create a helper function similar to `ofRadix` in "UnsignedReal.scala", then use it to define the various fields in `FloatingParsers`?

j-mie6 replied:
Probably? What does the Scala code do?

MMZK1526 replied:
According to my understanding of [the code](https://github.com/j-mie6/parsley/blob/staging%2F5.0/parsley%2Fshared%2Fsrc%2Fmain%2Fscala%2Fparsley%2Ftoken%2Fnumeric%2FUnsignedReal.scala), there's a `private def ofRadix(radix: Int, digit: Parsley[Char], endLabel: LabelConfig)` which does all the heavy work, and `_decimal`, `_hexadecimal` etc. all uses it.

j-mie6 replied:
Then go with that then :)

MMZK1526 commented on src/Text/Gigaparsec/Internal/Token/Numeric.hs:
```diff
@@ -371,23 +371,35 @@ mkSigned NumericDesc{..} !unsigned !err = IntegerParsers {
-{-type FloatingParsers :: *
-data FloatingParsers = FloatingParsers {}
+type FloatingParsers :: *
+data FloatingParsers = FloatingParsers { 
+  -- | Parse a single real number literal in decimal form (base 10).
+    decimalFloating :: Parsec Rational
+  -- | Parse a single real number literal in hexadecimal form (base 16).
+  , hexadecimalFloating :: Parsec Rational
+  -- | Parse a single real number literal in octal form (base 8).
+  , octalFloating :: Parsec Rational
+  -- | Parse a single real number literal in binary form (base 2).
+  , binaryFloating :: Parsec Rational
+  -- | Parse a single real number literal, 
+  -- which can be in many forms and bases depending on the configuration.
+  , numberFloating :: Parsec Rational
+  }
```
While I can use `DuplicateRecordFields`, it's quite unwieldy when actually using the duplicated field selectors in the same module as it complains about ambiguous names. Therefore I added the `-Floating` suffix in the field names.

Alternatively, we could
- Define `FloatingParsers` in a new module
- Use `DuplicateRecordFields` and use `getField @"decimal" xxx` instead of `decimal xxx` to avoid the ambiguity. But `getField` is not Safe Haskell
- Maybe there are other ways that I'm not aware of

j-mie6 replied:
There is a precedent set somewhere else for how to resolve the ambiguous names, I think? Just check around

MMZK1526 replied:
I'm not aware, please let me know 🙏🏼

j-mie6 replied:
I think I ran into overloaded field names for integer parsers no? Maybe that was their base names themselves which was later changed in parsley too

MMZK1526 replied:
Looks like there's no `DuplicateRecordFields` extension used in the codebase.

j-mie6 replied:
I mean I just added a suffix of some kind

MMZK1526 replied:
OK then, so all good here

MMZK1526 commented:
How do we translate the Scala `apply` override to Haskell? Is there a typeclass for that? 🙏🏼 

For example, all subclass of `PreventDotIsZeroConfig` has an implementation of `apply` which takes a boolean parser and produces a boolean parser. I tried to implement `PreventDotIsZeroConfig` as a normal Haskell `data`, which of course doesn't have the Scala syntax sugar of using it as a function after overriding an `apply` method.

j-mie6 commented:
That's codata vs data. Scala has codata (methods), you don't. As such, you just write a function on `PreventDotIsZeroConfig -> Parser ...` and pattern match

MMZK1526 committed:
feat: add real and combined (int or real) parsers

The implementation logics should largely match the scala ones, albeit to the best of effort : )

Using Rational as the result type, as recommended

MMZK1526 commented:
It's been a year :lol: but, here we go :)

j-mie6 commented:
Phew
